libkmod/libkmod-builtin.c

```diff
@@ -1,333 +1,185 @@
 }
 
-static bool kmod_builtin_iter_next(struct kmod_builtin_iter *iter)
+static ssize_t get_strings(struct kmod_builtin_info *info, const char *modname,
+				struct strbuf *buf)
 {
-	char *line,  *modname;
-	size_t linesz;
-	off_t pos, offset, modlen;
-
-	modname = NULL;
+	const size_t modlen = strlen(modname);
+	ssize_t count = 0, n;
 
-	iter->nstrings = 0;
-	offset = pos = iter->next;
+	do {
+		char *dot, *line;
 
-	while (offset < iter->size) {
-		char *dot;
-		off_t len;
-
-		offset = get_string(iter, pos, &line, &linesz);
-		if (offset <= 0) {
-			if (offset)
-				ERR(iter->ctx, "get_string: %s\n", strerror(errno));
-			pos = iter->size;
+		n = get_string(info);
+		if (n == -1)
 			break;
-		}
 
+		line = info->buf;
 		dot = strchr(line, '.');
-		if (!dot) {
-			ERR(iter->ctx, "kmod_builtin_iter_next: unexpected string without modname prefix\n");
-			pos = iter->size;
-			break;
+		if (dot == NULL) {
+			count = -EINVAL;
+			ERR(info->ctx, "get_strings: "
+			    "unexpected string without modname prefix\n");
+			return count;
 		}
-
-		len = dot - line;
-
-		if (!modname) {
-			modname = strdup(line);
-			modlen = len;
-		} else if (modlen != len || strncmp(modname, line, len)) {
+		if (strncmp(line, modname, modlen) ||
+		    line[modlen] != '.') {
+			if (count == 0)
+				continue;
 			break;
 		}
+		if (!strbuf_pushchars(buf, dot + 1) ||
+		    !strbuf_pushchar(buf, '\0')) {
+			count = -errno;
+			ERR(info->ctx, "get_strings: "
+			    "failed to append modinfo string\n");
+			return count;
+		}
+		if (count == INTPTR_MAX) {
+			count = -ENOMEM;
+			ERR(info->ctx, "get_strings: "
+			    "too many modinfo strings encountered\n");
+			return count;
+		}
+		count++;
+	} while (n != -1);
 
-		iter->nstrings++;
-		pos = offset;
+	if (n == -1 && !feof(info->fp)) {
+		count = -errno;
+		ERR(info->ctx, "get_string: %s\n", strerror(errno));
 	}
 
-	iter->pos = iter->next;
-	iter->next = pos;
-
-	free(modname);
-
-	return (iter->pos < iter->size);
+	return count;
 }
 
-static bool kmod_builtin_iter_get_modname(struct kmod_builtin_iter *iter,
-				char modname[static PATH_MAX])
+static char **strbuf_to_vector(struct strbuf *buf, size_t count)
 {
-	int sv_errno;
-	char *line, *dot;
-	size_t linesz, len;
-	off_t offset;
-
-	if (iter->pos == iter->size) {
-		sv_errno = EINVAL;
-		goto fail;
-	}
-
-	line = NULL;
-
-	offset = get_string(iter, iter->pos, &line, &linesz);
-	if (offset <= 0) {
-		sv_errno = errno;
-		if (offset)
-			ERR(iter->ctx, "get_string: %s\n", strerror(errno));
-		goto fail;
-	}
+	char **vector;
+	char *s;
+	size_t n;
 
-	dot = strchr(line, '.');
-	if (!dot) {
-		sv_errno = EINVAL;
-		ERR(iter->ctx, "kmod_builtin_iter_get_modname: unexpected string without modname prefix\n");
-		goto fail;
+	if (SIZE_MAX / sizeof(char *) - 1 < count ||
+	    SIZE_MAX - buf->used < sizeof(char *) * (count + 1)) {
+		errno = ENOMEM;
+		return NULL;
 	}
 
-	len = dot - line;
+	vector = realloc(buf->bytes, buf->used + sizeof(char *) * (count + 1));
```

Comment:
It probably won't hurt but this file will have a "special" strbuf handling anyway, because it puts NUL bytes into it on purpose. If at all, I would expect the most negative feedback on this PR for doing exactly that. ;)